### ISSUE TITLE
Syncthing Security - Information Leakage - Relaying

### ISSUE THREAD

**Issue by PascalHeynmoeller** [freedombox/Plinth#1120]
@JosephKiranBabu mybe you could fix this also?

https://docs.syncthing.net/users/relaying.html

"Relaying is **enabled** by default but will only be used if two devices are unable to communicate directly with each other. When connected via a relay, Syncthing will periodically retry a direct connection and, if one is established, stop communicating via the relay."
...
Security
..."In that respect it is similar to a discovery server."
Relaying should also be **disabled** by default on Freedombox
#1115

**Comment by JosephKiranBabu** [comment]
This would basically make Syncthing useless. The whole use case of syncthing is as an alternative to the centralized cloud services, with similar usability. At least that's how I use it. As soon as I click a picture, Apple Cloud or Google Cloud or NextCloud can instantly upload it to the server. So can Syncthing. We'll be taking away one of the core features of Syncthing by implementing this.

**Comment by PascalHeynmoeller** [comment]
@JosephKiranBabu 
Localy it works so it isn't useless in many cases!
I love nextcloud. Also Nextcloud doesn't use UID's which compromise the users.
Everyone is free to change the settings who doesn't need anonymity when they're as default disabled.
You can enable them in the Syncthing Web-GUI with some clicks.
I wish you would fix it as it fit our goals at https://freedomboxfoundation.org/goals/index.en.html
and https://freedomboxfoundation.org/learn/

**Comment by PascalHeynmoeller** [comment]
https://wiki.debian.org/FreedomBox/ProgressCalls/2017-07-23
"...
Syncthing

    Study relay nodes and provide checkbox to enable relay node on each FreedomBox. (3 points)
        Spike on how to setup a relay node.
        Check if it is a privacy violation. 
..."
Yes it is a privacy violation.

**Comment by JosephKiranBabu** [comment]
@PascalHeynmoeller 

### Privacy, anonymity and FreedomBox project goals
The FreedomBox project promises privacy, not perfect anonymity.
Anonymous publication is listed as one of the project goals. But we don't have the applications available for it yet, except for access using tor onion address of course.
The only truly anonymous service on the FreedomBox is **tor**, in my opinion. BitTorrent applications are not anonymous since they use trackers. Having a public ip address for your FreedomBox or using a Dynamic DNS violates anonymity as well. If you access your NextCloud instance using anything other than a _tor_ onion url, that's a UID that can be tracked too. Currently, accessing all of your services through _tor_, if they support it, is the best anonymity option we have. Syncthing unfortunately doesn't work like that, though it can be forced to. 
The most important way FreedomBox protects users' privacy is by encrypting data in motion and storing data in a location that the user controls. 
We should make it explicit that a particular application doesn't give anonymity guarantees. See #933.
Till that feature is implemented, it should either be mentioned in the wiki or description. 
AFAIK, FreedomBox can protect you from dragnet surveillance, not if some three-letter organization is directly targeting you. 

### Syncthing
Syncthing should be viewed as an alternative to Dropbox where you store the data in your personal cloud, instead of a company's cloud. It's not meant to be an anonymous file sharing system. We should add other anonymous file-sharing systems to FreedomBox, like _onionshare_. Tweaking Syncthing into something it is not will seriously damage its usability.

#### Syncthing relays
I thought of enabling Syncthing relays on FreedomBox as a means of decentralizing the relay system and also make it more fault-tolerant. We already have _tor_ relaying option on FreedomBox. 
To be a relay, one must have a public ip address and be listed on a central list of relay servers. This will expose the relay's physical location as well. It's a conscious risk that the host of the FreedomBox  should be willing to take. Every website we access on the internet as clients knows our client ip address. I expect most FreedomBox users to have some kind of public facing ip or DNS for their FreedomBox to be able to access it from outside the LAN. That said, I think a user that wants to be really anonymous should know what services they can use via _tor_. I will add that to #933.

**Comment by jvalleroy** [comment]
Currently, the Syncthing page in Plinth doesn't mention relay, or that it will register with a 3rd party. Shouldn't this be added, so the FreedomBox owner is aware of it? (Also applies to the Tor page, which only mentions it after install is completed.)

**Comment by JosephKiranBabu** [comment]
@jvalleroy Yes, the FreedomBox owner should be definitely aware of this. Shall we start by adding this information to the description, just like we did for Transmission? This information may be conveyed with better visual markers in #933.

**Comment by JoKeyser** [comment]
I agree with @JosephKiranBabu point on distinguishing anonymity vs privacy, so the default makes sense to me as it is (provided it's clear). I still understand @PascalHeynmoeller 's point that maybe some users want to sync only in the local network, so adding the (non-default) option would make sense.

**Comment by PascalHeynmoeller** [comment]
Tor Relay by default enabled is also an issue from my point of view.

When an Service is installed in Debian most services don't do anything without being enabled and started sometimes also without being configured.

**This is about Freedom of choice**

In Debian users can configure a service in an secure manner before they start the service after installation. At the moment we take this freedom of choice from the users and decide for them. These decisions should be made in a wise manner and result in an secure default configuration as expected by the users which are interested in not giving up their freedom.
Luckily TOR options can be disabled in plinth but this not the way I would expect the behavior of an Debian based system. The right behavior would be an disabled service which can be configured (enabling all the things needed) and after that start (enable). Users should have the control.
Also there are different use cases and circumstances for the use of Freedombox which should be respected. One of this use case is leaving the cloud using Freedombox as Intranetserver.
Syncthing is an tool for synchronisation but you can use it for filesharing. This decision should be made by the owner of the box not by us how to use it. Thats why in my opinion Relay and Global Dir should be disabled by default (same with TOR Relay), two checkboxes are needed for enabling them.
This would be a privacy friendly solution with the freedom of choice.

@JosephKiranBabu if you still think "default disabled" is wrong, give users the choice to disable these two options in Plinth with some toggle buttons or at least we should point them to the admin webinterface setting section for disabling this behavior.

**Comment by JoKeyser** [comment]
@PascalHeynmoeller I agree with giving the users the choice in this case. In general I agree with @JosephKiranBabu 's leaning towards usefulness. To me, FreedomBox is not "AnonymityBox" as *first* goal, but it's more important to first provide as-easy-as-possible useful services. We may have to deviate somewhat from Debian's very conservative approach to achieve that ease of use. With every additional click, we will lose a fraction of potential users. Of course that's a tricky balance to strike, so it's great (in fact, vital) to have discussions like these.
Could somebody maybe clarify (for future readers, and to users) what kind of information is leaked exactly in this case? What can be inferred from these data?

**Comment by JoKeyser** [comment]
Oh, and in this case, the defaults used in FreedomBox and Debian are actually the same, aren't they? I see no changes of settings in https://github.com/freedombox/Plinth/tree/master/plinth/modules/syncthing ... so that part of the conversation isn't a concern. (I also should clarify that I agree with Debian's stance 99% of the time so far, but I maintain that the goals of our sub-project isn't the same, generally speaking.)

**Comment by PascalHeynmoeller** [comment]
@JoKeyser The Problem is an other, as mentioned in my first post in this thread, you can also have a look at #1115 where it is explained.
They are enabled. You are right, the defaults are the same (have look at #1115 for deeper understanding what happened) but no, the point of the conversation is that Debian gives me the choice to configure services **before** they are enabled and started. "User Controlled"
This is an generally different behavior if you install software via Plinth as James Valleroy mentioned above. Plinth is our benefit so it is ok to do it the way we do it. (first Plinth then the User)
As written before we should be more careful in making decisions for users in Plinth.
In my opinion should users decide which particular feature of an Service he want to use in his/her case before the service is enabled and started. Especially if they can potentially violate users privacy. This suppose an disabled Service when it is installed via Plinth. Then users have the chance to toggle on/off features before they enable the Service. #933 Also there should be warnings on single features which helps users to decide if a feature is against his or her interests. (Power to the people)
When I want to synchronize files between devices in my local Network I should be empowered to do it without violating my privacy the same with TOR when I just want to use it for downloading Debian packages anonymous.
In between it became for me an fundamental issue on how we violate privacy by default and infantilise users by making choices for them, it feels like heading in the wrong direction ,sadly. I hope this direction will be corrected in future.
Still like to discuss this general issue with more people.

**Comment by mray** [comment]
I second @PascalHeynmoeller s concerns. If installing an application *automatically* starts phoning home to some third party server that is a problem in any case. For the same reasons we would not be ok with setting up a third party DNS service by *default* – even if it would solve *A LOT* of our problems.

Presenting a prominent option to enable global reachability after installation should work for everybody.

**Comment by JoKeyser** [comment]
@PascalHeynmoeller , @mray First off, I think we're all pretty much on the same page, and technically only talking about nuances; we all agree that there should be an option to enable/disable that option. So either a) an option to *enable after* starting the service or b) an option to *disable* the service *before* starting it would give the user the control we agree they should have.
I'm favoring b), in line with Debian, because defaults are very strong and should fit the use-case of most users. I'm concerned that reading about an option is too much friction for our main audience in (relatively) low-risk scenarios, who will simply go elsewhere commercial cloud, which offers that ease of use. In high-risk scenarios (journalist in bad regime), it's not too much to ask to read a bit more, and disable the default. (FreedomBox isn't designed for high-risk anyways, at least not primarily.) If we agree, then that approach of setting-defaults, but disabling until manual start should probably be the same for all/most services?


Apart from that, I'd like to comment that I disagree with some of your verbiage because it strikes me too extreme and misleading.
I disagree that providing (sensible) defaults and reducing the choices are somehow "infantilising" the users. Indeed, it's the project's goal to do exactly that, and it is what's *necessary* to make the software accessible to a wide non-technical audience, which is the primary goal.
I disagree that the option in question will "automatically starts phoning home" - it's initiated by the user by setting up the devices and choosing folders to sync, and the 3rd party is only relaying encrypted data. I think that's comparable to sending an encrypted email through servers you don't control.
I also don't really understand what the scenario should be where knowledge of 3rd parties that you're (probably) running Syncthing constitutes a "violation of privacy", except in the narrow sense of losing some anonymity (as you say it's similar to running a Tor relay). If so, let's use that word instead, for the sake of remaining precise.

TL;DR, I think enabling this option is vastly more useful than it is a threat to privacy, and hence should remain the default. Again, I agree that there should be an option to disable it.

**Comment by mray** [comment]
Why not c) option to *enable* before the service starts?

I usually am a strong proponent of having less choice and having a frictionless experience, but since this matter has strong ties to the **Freedom** in **Freedom**Box I feel we are in a delicate situation here and should not set a precedent that questions our goals and values or how we go about them. If people are not willing to do *some extra work* they wouldn't go through the hassle of getting a FreedomBox running in the first place. Conveniences like that are ok to be one click further *away* in my opinion.

Lets put it at the fingertips for people to compromise, but not make it a default.

**Comment by jvalleroy** [comment]
We already have some services with an extra setup step after installation. Could we do something like this?

Syncthing packages installed. Please select an option to complete setup:
- [X] **Use recommended settings**
        Syncthing will register with a 3rd party service, to provide additional functionality. Any data sent through this service will be encrypted.
- [ ] **Prefer privacy over functionality**
        Disable registration with 3rd party service. This may limit availability of some features.

[Finish setup]

**Comment by JoKeyser** [comment]
I agree that it's a delicate choice - I can live with either option. Option c) feels redundant to me, and a) would be better in that case: if it's disabled by default anyways, there's no point in waiting to start the service.
I think with b) we'd prompt users with enough caution to think about the default. If we change the default, we would deviate from the expected behavior from all other devices that have it installed. That's why this discussion seems suitable to have upstream.

**Comment by JoKeyser** [comment]
I can live with that solution and interpret this purely as a documentation issue. Arguably users *have* control by simply never installing this application. I trust @JosephKiranBabu assessment that Syncthing is useless without these options. Other apps should be used for anonymous or local-only file synchronization.

**Comment by PascalHeynmoeller** [comment]
@JoKeyser I'm sry for picking to strong words from my translator. I'm not a native English speaker, writing in an foreign language costs a lot of time and leads sometimes to misunderstanding. So it's not easy for me to discuss. I am glad that some got what the real issue is.

I'm also fine with option b.) .

I still find it hard to say its general useless but I'm with you @JosephKiranBabu and agree that every service with this general problematic should be treated in the same way for better privacy. We can reach this by using the patter from the suggestion @JoKeyser and @jvalleroy  made. 
I feel sry for you that you think someone is trying to hammer it into something it is not. Maybe it's because you think it is a Dropbox replacement like others do but that's not what it is originally made for. It's an replacement for the proprietary https://en.wikipedia.org/wiki/BitTorrent_Sync now Resilio Sync advertised as sync solution in general.

Some views changed in my mind the last days, turned it into a bigger problem as it was in the beginning.
Jumping from one issue to the next, being pointed to the fact that there are more services with the same problematic, while reading the Syncthing Handbook new aspects appeared also during these discussions and listing to Eben Moglen's presentations while thinking about it.

For example, if the discovery server would be the freedombox and also the relay, nothing would go out by default. I just found out in the last minutes that this also can be a solution which could be possible. Both (discover and relay sever) are available in Debian exactly for this case to decentralize the service.
So Debian default is carefully selected when you host both services by your self. (But toggles are still needed when installed by plinth).

For me Freedombox advertised as on https://freedomboxfoundation.org/learn/ with strong anonymity features is one of the key advantages and distinct it from other projects which provides one-click selfhost solutions.

@JosephKiranBabu thank you for your invitation to the wiki, I've contributed to the wiki in the past and try to write a security section for Syncthing pointing to the handbook an writing a hint about how to disable GlobalDir and Relay till we have some toggles as shortcut for doing it.

**Comment by mray** [comment]
@JosephKiranBabu thanks for laying out your point so patiently. I didn't have a complete technical understanding as it turns out. I'm ok b), too.

**Comment by testbird** [comment]
The perfect default syncthing setup provides per user syncthing instances together with an own private relay server
https://forum.syncthing.net/t/private-relay-server-with-client-authentication/8435
and an own discovery server
https://docs.syncthing.net/users/stdiscosrv.html

**Comment by testbird** [comment]
... the users can then point syncthing on their devices to the freedombox syncthing (including the discovery and relay server)

**Comment by testbird** [comment]
BTW the wording in the mockup above may benefit from more precision, e.g. instead of marketing like talk of "recommended settings" for using the central servers, maybe: 

 [*] let this freedombox provide own discovery and relay services (increases independence but also exposure of potential vulnerabilities)

**Comment by JoKeyser** [comment]
@testbird thank you, using FreedomBox as your own relay and discovery makes total sense to me. Do you know if/how well this works through NAT? (Even if it doesn't it would make sense for boxes without NAT.) Upstream is an open issue to allow only private connections to the relay instance, https://github.com/syncthing/syncthing/issues/3987. Seems like this would be a precursor before we can meaningfully add this? (The point being that running your own relay is to enhance privacy.)
Maybe it would be good to split discovery / relay / multiple users into separate, new issues, and close this one?

**Comment by testbird** [comment]
An own relay server allows private connections without external relays.
Disallowing unknown clients seems an aditional security measure to me.

**Comment by PascalHeynmoeller** [comment]
@testbird after digging deeper into the docs I came to the same conclusion like you.

Relaysrv and discosrv, both are needed and it's not easy to run syncthing and discosrv on the same machine at the same time because both use the same port. Also DynDNS is needed for dynamic IP's.
@JosephKiranBabu the conclusion is Syncthings infrastructure rely on Client-Server (DNS) for discovery and relay. Without DNS Syncthing can't find each other (initially) when they're on different Networks because they can't register to an relay or discosrv without DNS. So Syncthing is only an Client-Server less service when it is used on the same network segment using local discovery in every other case it depends on DNS client and servers to connect to discosrv and relaysrv to find each other. Without  telephone book you can't make long-distance calls. It's not like a trackerless bittorent (sadly). [Use decentralized discovery (DHT)](https://github.com/syncthing/syncthing/issues/3388)

Referring to documentation doesn't resolve this issue because it is still there.
Option b.) would empower the user to solve this issue for him self with a single click, telling syncthing not to talk to strangers outside the own network. (Syncthing can also limit access from an IP-Range in the setting section to allow only clients from the VPN for example).

Decentralizing by deploying relaysrv and discosrv would be the best approach. Additional to option b.).

@testbird, @JoKeyser may someone of you like to open issues for deploying syncthing-relaysrv & -discosrv on Freedombox as an alternative for decentralization and privacy and attach #1120 and #1115 to solve them. I don't have the time at the moment.

**Comment by JosephKiranBabu** [comment]
@PascalHeynmoeller There is hardly any application on FreedomBox that is ***absolutely*** decentralized (or anonymous). Centralized DNS is a larger problem which I hope would be addressed by some other project, probably by using a blockchain.

The one task that I can agree without any objection is this:
Update wiki and description to make it clear that Syncthing is not anonymous and uses centralized infrastructure.

Since a user will use multiple Syncthing instances on multiple devices and we control only one device, we cannot automatically set the relay server and discovery server to the one on the user's FreedomBox device. That said, we should still help run relay servers and discovery servers to decentralize the Syncthing infrastructure.

> b) an option to disable it before the service starts would give the user the control we agree they should have

This would have to be done for most of the apps on FreedomBox. We have to make this a framework-level feature. 
To easily integrate with our current installation process, we can add a collapsible section above the "Install" button with checkboxes like in @jvalleroy 's mockup. If the user made a bad decision about this, they should be able to change the settings after installation too. So, the same section would also be included in the service page.

I will create issues for the above mentioned tasks.

**Comment by testbird** [comment]
Franky, I find mockup above kind of confusing because it does neither what will done, what this is for, nor what the consequences are.

Not sure if I understood everything well yet, still here is a draft:

Example:
- [x] **Register FreedomBox syncthing instances with public discovery servers** (default: yes)
    ```
    This allows other syncthing nodes to find and connect by only knowing
    the node ID (look up the server address on 3rd party servers).
    ```
- [x] **Let FreedomBox provide own discovery and relay services** (default: yes)
    ```
    This allows improved privacy by configuring other syncthing nodes to
    only connect to your freedombox.

        The advanced settings of other nodes may then be configured:

           discovery server: fbx.domain:xxxx
           relay server: fbx.domain: yyyy
   ```

**Comment by JoKeyser** [comment]
@PascalHeynmoeller I like the additional info in your mockup. For the first entry, maybe even more information (by a link?) could be included, about which public servers are being used per default.
The second might be better to split into discovery and relay services, and could then be treated in their respective issue, e.g. #1136 for discovery ... (It's good to split up the tasks like that into their smallest useful chunks, to encourage implementation.)
